File: dotnet/CardsProject/Domain/Content/Stream.cs
using CardsProject.Domain.Tournaments;
using CardsProject.Domain.Players;
using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;

namespace CardsProject.Domain.Content;

public enum StreamStatusType
{
    Scheduled,
    Live,
    Ended
}

public enum StreamPlatformType
{
    Twitch,
    YouTube,
    KickStream,
    Platform
}

public enum StreamLanguageType
{
    EN,
    DE,
    FR,
    IT,
    ES,
    JP,
    PT
}

public class Stream : IValidatableObject
{
    public int Id { get; set; }

    public string Title { get; set; } = "";
    public string StreamUrl { get; set; } = "";
    public StreamStatusType Status { get; set; }
    public StreamPlatformType Platform { get; set; }
    public StreamLanguageType Language { get; set; }
    public bool IsOfficial { get; set; } = false;
    public int ViewerCountPeak { get; set; } = 0;
    public DateTime? ScheduledStart { get; set; } = null;
    public DateTime? ActualStart { get; set; } = null;
    public DateTime? EndedAt { get; set; } = null;
    public string? VodUrl { get; set; }

    public int? TournamentId { get; set; }
    [ForeignKey(nameof(TournamentId))]
    public Tournament? Tournament { get; set; }
    public int? StreamerId { get; set; }
    [ForeignKey(nameof(StreamerId))]
    public Player? Streamer { get; set; }

    // Business operations

    public void GoLive()
    {
        // TODO: implement go_live
    }

    public void End()
    {
        // TODO: implement end
    }

    public void UpdateViewerPeak(int count)
    {
        // TODO: implement update_viewer_peak
    }

    public int DurationMinutes()
    {
        // TODO: implement duration_minutes
        return default;
    }

    // ── Lifecycle state machine ──────────────────────────────────────
    private static readonly System.Collections.Generic.Dictionary<StreamStatusType, StreamStatusType[]> AllowedTransitions = new()
    {
        [StreamStatusType.Scheduled] = new[] { StreamStatusType.Live },
        [StreamStatusType.Live] = new[] { StreamStatusType.Ended }
    };

    public void AssertTransition(StreamStatusType to)
    {
        if (!AllowedTransitions.TryGetValue(Status, out var allowed) || !System.Array.Exists(allowed, s => s == to))
            throw new InvalidOperationException($"Transition {Status} -> {to} not allowed");
    }

    // ── Domain invariants (simple rules) ──────────────────────────────
    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
    {
        if (!( ViewerCountPeak >= 0 ))
            yield return new ValidationResult("Peak viewer count must not be negative", new[] { nameof(Id) });
    }
}

File: phoenix/lib/cards_project/content/stream.ex
defmodule CardsProject.Content.Stream do
  use Ecto.Schema
  import Ecto.Changeset

  schema "streams" do
    field :title, :string
    field :stream_url, :string
    field :status, :string
    field :platform, :string
    field :language, :string
    field :is_official, :boolean, default: false
    field :viewer_count_peak, :integer, default: 0
    field :scheduled_start, :naive_datetime
    field :actual_start, :naive_datetime
    field :ended_at, :naive_datetime
    field :vod_url, :string
    belongs_to :tournament, CardsProject.Tournaments.Tournament
    belongs_to :streamer, CardsProject.Players.Player

    timestamps()
  end

  @doc false
  def changeset(record, attrs) do
    record
    |> cast(attrs, [:title, :stream_url, :is_official, :viewer_count_peak, :scheduled_start, :status, :platform, :language, :actual_start, :ended_at, :vod_url, :tournament_id, :streamer_id])
    |> validate_required([:title, :stream_url, :is_official, :viewer_count_peak, :scheduled_start])
    |> validate_inclusion(:status, ["Scheduled", "Live", "Ended"])
    |> validate_inclusion(:platform, ["Twitch", "YouTube", "KickStream", "Platform"])
    |> validate_inclusion(:language, ["EN", "DE", "FR", "IT", "ES", "JP", "PT"])
    |> validate_number(:viewer_count_peak, greater_than_or_equal_to: 0, message: "Peak viewer count must not be negative")
    |> then(fn cs ->
      if not is_nil(get_field(cs, :actual_start)) and (get_field(cs, :status) != "Live") do
        Ecto.Changeset.add_error(cs, :status, "actual_start_requires_live_or_ended validation failed")
      else
        cs
      end
    end)
    |> then(fn cs ->
      if not is_nil(get_field(cs, :ended_at)) and (get_field(cs, :status) != "Ended") do
        Ecto.Changeset.add_error(cs, :status, "ended_at can only be set when stream status is Ended")
      else
        cs
      end
    end)
  end

  # ── Business operations ────────────────────────────────────────────

  def go_live(_record) do
    # TODO: implement Stream.go_live
    :ok
  end

  def end_action(_record) do
    # TODO: implement Stream.end_action
    :ok
  end

  def update_viewer_peak(_record, _count) do
    # TODO: implement Stream.update_viewer_peak
    :ok
  end

  def duration_minutes(_record) do
    # TODO: implement Stream.duration_minutes
    {:error, :not_implemented}
  end

  # ── Lifecycle state machine ─────────────────────────────────────────
  @allowed_transitions %{
    "Scheduled" => ["Live"],
    "Live" => ["Ended"]
  }

  def assert_transition(%__MODULE__{status: current}, to) do
    allowed = Map.get(@allowed_transitions, current, [])
    if to in allowed do
      :ok
    else
      {:error, "Transition #{current} -> #{to} not allowed"}
    end
  end
end
